_posts/2023-08-13-kafka-node-pools-introduction.md

```diff
@@ -0,0 +1,347 @@
+But they will be completely ignored when node pools are used.
+That way we make sure that the schema validation of the `Kafka` custom resources works well for the majority of Strimzi users who will have the feature gate disabled.
+Once the node pools move to beta and will be enabled by default, these fields will be made optional and it will not be required to set them anymore.
+
+#### Examples
+
+Let's look at some example.
+First, the `KafkaNodePool` resources:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: pool-a
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  roles:
+    - broker
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 200Gi
+        deleteClaim: false
+  resources:
+    requests:
+      memory: 16Gi
+      cpu: 1000m
+    limits:
+      memory: 16Gi
+      cpu: 2000m
+  jvmOptions:
+    -Xms: 4096m
+    -Xmx: 4096m
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: pool-b
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  roles:
+    - broker
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 1Ti
+        deleteClaim: false
+  resources:
+    requests:
+      memory: 32Gi
+      cpu: 2000m
+    limits:
+      memory: 32Gi
+      cpu: 4000m
+  jvmOptions:
+    -Xms: 8192m
+    -Xmx: 8192m
+```
+
+Above, you can see two `KafkaNodePools` named `pool-a` and `pool-b`.
+Each of them has a slightly different configuration.
+The nodes in `pool-b` are bigger then in `pool-a`, have more storage, and more memory used for the Java heap.
+Both node pools have the `strimzi.io/cluster` label pointing to a Kafka cluster named `my-cluster`.
+So we need to have a `Kafka` custom resource named `my-cluster` as well:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/node-pools: enabled
+spec:
+  kafka:
+    version: 3.5.1
+    # The replicas field is required by the Kafka CRD schema while the KafkaNodePools feature gate is in alpha phase.
+    # But it will be ignored when Kafka Node Pools are used
+    replicas: 3
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.5"
+    # The storage field is required by the Kafka CRD schema while the KafkaNodePools feature gate is in alpha phase.
+    # But it will be ignored when Kafka Node Pools are used
+    storage:
+      type: jbod
+      volumes:
```

Comment:
more technical question than related to the article itself but ... if the storage is ignored, how the operator behaves if I set `volumes: {}` or as an empty array? Does the Kube validation fail? Just curious because in the end you are specifying a storage size which is totally ignored so maybe just leaving it out provides more readability of the `Kafka` custom resource.

Reply:
I don't think you can set it to an empty object because, without the type, you have no class to decode it into. (`Storage` is abstract class and cannot be used)